test/correctness/extern_reorder_storage.cpp

```diff
@@ -0,0 +1,60 @@
+#include "Halide.h"
+#include <stdio.h>
+
+#ifdef _WIN32
+#define DLLEXPORT __declspec(dllexport)
+#else
+#define DLLEXPORT
+#endif
+
+// An extern stage that translates.
+extern "C" DLLEXPORT int copy_and_check_strides(halide_buffer_t *in, halide_buffer_t *out) {
+    if (in->is_bounds_query()) {
+        for (int i = 0; i < 2; i++) {
+	    in->dim[i].min = out->dim[i].min;
+	    in->dim[i].extent = out->dim[i].extent;
+	}
+    } else if (!out->is_bounds_query()) {
+        // Check that the storage has been reordered.
+        assert(out->dim[0].stride > out->dim[1].stride);
+        Halide::Runtime::Buffer<uint8_t> out_buf(*out);
+        out_buf.copy_from(Halide::Runtime::Buffer<uint8_t>(*in));
+    }
+
+    return 0;
+}
+
+using namespace Halide;
+
+int main(int argc, char **argv) {
+    Var x, y;
+
+    const int W = 30, H = 20;
+    Buffer<uint8_t> input_buffer(W, H);
+    for (int i = 0; i < H; i++) {
+        for (int j = 0; j < W; j++) {
+            input_buffer(j, i) = i + j;
+        }
+    }
+
+    // Define a pipeline that uses an input image in an extern stage
+    // only and do bounds queries.
+    ImageParam input(UInt(8), 2);
+    Func f, g;
+
+    f.define_extern("copy_and_check_strides", {input}, UInt(8), {x, y});
```

Comment:
If we're going to add an overload, it might be worth reconsidering define_extern syntax in general. What's the ideal syntax here? We could for example make a new object to represent the RHS of an extern definition and give FuncRef an operator= that accepts it:

f(x, y) = Extern<uint8_t>("copy_and_check_strides", input) ?

Reply:
Agreed that it's a good excuse to revisit the syntax. @abadams strawman looks like a good starting point.